### ISSUE TITLE
Cannot overfit net on a small training set

### ISSUE THREAD

**Issue by sptom** [milesial/Pytorch-UNet#165]
Hi there,
First of all, thank you for this code!
My first run with the network was unsuccessful, so I tried to make a quick sanity check.
It is my understanding that a "healthy" neural net should easily overfit the training data when few examples are given. It should quickly learn to classify them with 100% accuracy by simply "memorising" the images.
So I tried to overfit the net for 38 images (35, actually, because 3 are used as validation set) from the Pascal VOC database.
![image](https://user-images.githubusercontent.com/12126222/80083549-6704a880-855e-11ea-85a7-b759d435f48a.png)

I use binary black=background white=object masks.
![image](https://user-images.githubusercontent.com/12126222/80084431-88b25f80-855f-11ea-9f93-b94f07644a61.png)

Running training for 100 epochs, I still get very poor results - the training loss fluctuates rather heavily around 0.5, as can be seen from the tensor-board plot and the output to screen:
![image](https://user-images.githubusercontent.com/12126222/80085033-60773080-8560-11ea-8b0a-db799da434e9.png)

![image](https://user-images.githubusercontent.com/12126222/80105404-3bdf8080-8581-11ea-8cce-1f7448081a74.png)

I'm quite certain that this behaviour is quite irregular and that I'm missing something. Do you have an idea what might that be?

The results on the very same training set don't make any sense:
![image](https://user-images.githubusercontent.com/12126222/80105753-aabcd980-8581-11ea-8ef1-5438e4bd2281.png)

**Comment by milesial** [comment]
Thank you for the detailed explaination of your problem. From the loss value and the loss plot it seems that your learning rate is way too high, have you tried lowering it? Divide it by 10 first maybe. 

Also for a sanity check, you can try with even less images, like 1 or 2, the training will be faster.

**Comment by juliagong** [comment]
Hi @milesial and @sptom! I'm having this issue of failing to overfit as well, and I'm using 1e-4 as my learning rate and a training set of just a single image. I can't seem to find anything irregular in the training code, but if more than one of us is having this issue, I wonder if there's something we're overlooking? Any insight would be appreciated!

**Comment by milesial** [comment]
I think your learning rate is too high. For the full carvana dataset I used 2e-6 as a LR

**Comment by juliagong** [comment]
Thanks; I unfortunately have also tried learning rates on the order of 1e-6 and got the same results. I also disabled all augmentation, normalization, and other regularization so that it's exactly the steps that you used. For some reason, the model isn't overfitting to the one-image dataset and either gives nonsensical segmentations or converges to weights being all 0. Have you had this issue? Do you have any insights? Thanks!

**Comment by sptom** [comment]
Thanks for the reply Alex,
I also tried running the code with various lower orders of magnitude for the learning rate. I also tried using different optimizers, but I have to agree with Julia here, it did not help:
The phenomenon that I see is that the lower I take the LR, the quicker I get the loss to converge to 0.7 and stay there, for some reason. 
![image](https://user-images.githubusercontent.com/12126222/80761802-eccaba00-8b43-11ea-9096-cb2c73eda5c9.png)

In some of the runs I got somewhat indicative results for the masks, but it's hardly something that could be considered 'overfitting'.
![image](https://user-images.githubusercontent.com/12126222/80761829-f9e7a900-8b43-11ea-8e7d-64f8dc777611.png)

Could it be that the loss funciton is unstable? Or perhaps you added some alteration in a recent update?

Thanks a lot, 
Tom

**Comment by milesial** [comment]
There was a lot of small tweaks in recent commits, but nothing that should affect the convergence I think. Are you using transposed conv or the bilinear route (default)?
The loss is just the cross entropy so it should be pretty stable.

Do you both work on the same dataset? Have you tried with an image form the Carvana dataset?

This problem is very strange. If you feed it 100 images, does it learn something?

**Comment by juliagong** [comment]
I'm using the bilinear route. I don't think we're using the same dataset, but we have the same issue. I've tried feeding it 50-100 images in training and it doesn't learn properly; it stays at around 0.7 loss.

**Comment by juliagong** [comment]
Update: I also tried transposed convolutions and they are not working either. It's such a strange issue and I don't think I've ever encountered something like this before.

**Comment by juliagong** [comment]
@milesial I spent today debugging once again by rewriting the entire training pipeline from scratch and testing on incrementally more meaningful sets of data, and ended up finding the problem! It was very sneaky.

Your `train.py` is actually fine, but in `eval.py`, `net.eval()` was called. However, `net.train()` was called only at the beginning of each epoch, while `net.eval()` was called every batch. So there was only meaningful training going on for one batch per epoch - no wonder! :)

Thanks for your help and quick response on this problem. I hope this fixes @sptom's issue as well.

**Comment by sptom** [comment]
Oh, Wow, awesome @juliagong ! that sounds really sneaky! I didn't notice that the net.train() was put epoch and not in the batch loop...
However, I tried to solve this, first by putting net.train() into the batch loop, which didn't help unfortunately. I then also tried to put the code in eval.py into with torch.no_grad(): instead of using net.eval(), but I didn't notice any significant effect by that.
Could you say how did you resolve the issue, and also - which parameters did you use afterwards?
Was the overfitting accurate? How many epochs did it take to converge to effectively zero loss?

**Comment by phper5** [comment]
> @milesial I spent today debugging once again by rewriting the entire training pipeline from scratch and testing on incrementally more meaningful sets of data, and ended up finding the problem! It was very sneaky.
> 
> Your `train.py` is actually fine, but in `eval.py`, `net.eval()` was called. However, `net.train()` was called only at the beginning of each epoch, while `net.eval()` was called every batch. So there was only meaningful training going on for one batch per epoch - no wonder! :)
> 
> Thanks for your help and quick response on this problem. I hope this fixes @sptom's issue as well.

i think the real train code is `masks_pred = net(imgs)`  and then count the loss  and `loss.backward()`  so every batch it has traind, even net.train() was called only at the beginning of each epoch.  am i wrong? thanks

**Comment by sptom** [comment]
@phper5, for masks_pred = net(imgs) you are right,
But the script does calculate the accuracy on the test set several times in an epoch using eval.py

**Comment by milesial** [comment]
@juliagong Thanks for your investigation ! It is indeed a big mistake that is there from a long time ago and needs a fix. But do the `train` and `eval` methods of the net module affect something else than BatchNorms here?
When you say there is no meaningful training, I'm not sure, since there is still gradient updates on other layers, it's just the BatchNorms that are broken (?)

Thanks to all of you for participating in this

**Comment by phper5** [comment]
> @phper5, for masks_pred = net(imgs) you are right,
> But the script does calculate the accuracy on the test set several times in an epoch using eval.py

Yes  you are right. sorry I didn't look carefully

**Comment by sptom** [comment]
> Oh, Wow, awesome @juliagong ! that sounds really sneaky! I didn't notice that the net.train() was put in the epoch and not in the batch loop...
> However, I tried to solve this, first by putting net.train() into the batch loop, which didn't help unfortunately. I then also tried to put the code in eval.py into with torch.no_grad(): instead of using net.eval(), but I didn't notice any significant effect by that.
> Could you say how did you resolve the issue, and also - which parameters did you use afterwards?
> Was the overfitting accurate? How many epochs did it take to converge to effectively zero loss?

@juliagong could you please share how did you resolve the issue? Because I did the above and it doesn't seem to help, still converging to 0.6 loss

**Comment by gboy2019** [comment]
> > Oh, Wow, awesome @juliagong ! that sounds really sneaky! I didn't notice that the net.train() was put in the epoch and not in the batch loop...
> > However, I tried to solve this, first by putting net.train() into the batch loop, which didn't help unfortunately. I then also tried to put the code in eval.py into with torch.no_grad(): instead of using net.eval(), but I didn't notice any significant effect by that.
> > Could you say how did you resolve the issue, and also - which parameters did you use afterwards?
> > Was the overfitting accurate? How many epochs did it take to converge to effectively zero loss?
> 
> @juliagong could you please share how did you resolve the issue? Because I did the above and it doesn't seem to help, still converging to 0.6 loss

me too，sometimes 0.7，sometimes 0.6，sometimes 0.8！so how fix this issue？

**Comment by ProfessorHuang** [comment]
> @milesial I spent today debugging once again by rewriting the entire training pipeline from scratch and testing on incrementally more meaningful sets of data, and ended up finding the problem! It was very sneaky.
> 
> Your `train.py` is actually fine, but in `eval.py`, `net.eval()` was called. However, `net.train()` was called only at the beginning of each epoch, while `net.eval()` was called every batch. So there was only meaningful training going on for one batch per epoch - no wonder! :)
> 
> Thanks for your help and quick response on this problem. I hope this fixes @sptom's issue as well.

Invaluable finding! Thank you very much. milesial's code is pretty, I think I can understand it, but I can't figure out Why it doesn't work on my dataset. Today, When I changed the code as you said, everything becomes good.

**Comment by milesial** [comment]
Hi all, I modified to switch back to train mode in https://github.com/milesial/Pytorch-UNet/commit/773ef215d41c1f36dc0ed4159c12df89d792fbc3

**Comment by sptom** [comment]
Thanks, @milesial  for your update.
However, I'm sorry to say that this **did not** resolve the issue.
Loss is still stuck around 0.6
I tried reduce the dataset to 11 copies of a single image, and the loss is now stock on 0.3:

![image](https://user-images.githubusercontent.com/12126222/81119982-7f000300-8f34-11ea-9098-89fbe646b18f.png)

Were you able to overfit a set of 10 different images? If so, how many epochs did it take, and which which parameters?

@ProfessorHuang , What exactly did you change in the code?

**Comment by gboy2019** [comment]
> Thanks, @milesial for your update.
> However, I'm sorry to say that this **did not** resolve the issue.
> Loss is still stuck around 0.6
> I tried reducing the dataset to 11 copies of a single image, and the loss is now stuck on 0.3:
> 
> ![image](https://user-images.githubusercontent.com/12126222/81119982-7f000300-8f34-11ea-9098-89fbe646b18f.png)
> 
> ![image](https://user-images.githubusercontent.com/12126222/81121139-b1aafb00-8f36-11ea-98f2-a330eb34c85d.png)
> The above is the result after running 15 epochs on 11 copies of the same image!
> 
> Were you able to overfit a set of 10 different images? If so, how many epochs did it take, and with which parameters?
> 
> @ProfessorHuang , What exactly did you change in the code?

could you share your code？loss of mine is 1e+3 sometimes，very terrible ，so show your code？

**Comment by shilei2403** [comment]
> > @milesial I spent today debugging once again by rewriting the entire training pipeline from scratch and testing on incrementally more meaningful sets of data, and ended up finding the problem! It was very sneaky.
> > Your `train.py` is actually fine, but in `eval.py`, `net.eval()` was called. However, `net.train()` was called only at the beginning of each epoch, while `net.eval()` was called every batch. So there was only meaningful training going on for one batch per epoch - no wonder! :)
> > Thanks for your help and quick response on this problem. I hope this fixes @sptom's issue as well.
> 
> Invaluable finding! Thank you very much. milesial's code is pretty, I think I can understand it, but I can't figure out Why it doesn't work on my dataset. Today, When I changed the code as you said, everything becomes good.

how do you change the code ?could you show the details?

**Comment by karlita101** [comment]
Hi there, I am also wondering if there's been any updates or willingness to share how they've overcame this issue.

Very much appreciated! 

@juliagong @ProfessorHuang

**Comment by Li-Wei-NCKU** [comment]
> Hi all, I modified to switch back to train mode in [773ef21](https://github.com/milesial/Pytorch-UNet/commit/773ef215d41c1f36dc0ed4159c12df89d792fbc3)

The modify makes sense to me logically.
However, the loss is still stuck and the model couldn't overfit on a small dataset with only 20 images.
Is there any kind suggestion? @milesial @juliagong @ProfessorHuang

**Comment by AJSVB** [comment]
I might be wrong, but I had similar issue, and reducing weight decay and momentum helped me overfitting

**Comment by rgkannan676** [comment]
Hi,

For small datasets, reducing the evaluation frequency reduced the training loss for me. This is will avoid learning rate becoming small after few steps.

Example. 
# Evaluation round
if global_step % (n_train // (0.25 * batch_size)) == 0:

**Comment by Flyingdog-Huang** [comment]
> Hi,
> 
> For small datasets, reducing the evaluation frequency reduced the training loss for me. This is will avoid learning rate becoming small value after few steps.
> 
> Example.
> 
> ```
> # Evaluation round
> if global_step % (n_train // (0.25 * batch_size)) == 0:
> ```

@rgkannan676  Thanks a lot ,this way helps the loss value becomes normal in small dataset for me ,I want to know why it can avoid LR becoming small after few steps, and now I will think how can reduce the loss shoking like this

![微信图片_20210911221748](https://user-images.githubusercontent.com/43697342/132950921-23a80676-e8a8-4cc3-971c-b98807491908.png)

hope can recieve your reply, thanks again @rgkannan676

**Comment by Flyingdog-Huang** [comment]
and also I find the reason ,the author has added 'net.train()' in evaluate.py, but it did not work, the model is still at 'eval()', it is not changed in training, so the model was trained one time pre epoch, the result likes this:
![微信图片_20210911221748](https://user-images.githubusercontent.com/43697342/132982189-147e3960-7c84-4054-89a6-7c875c3a9df0.png)

and I change the model to 'train()' after evaluation round, and I think it works, the result likes this:
![image](https://user-images.githubusercontent.com/43697342/132982310-f29b7523-48cb-4090-840a-c4989fa39c8d.png)

**Comment by Flyingdog-Huang** [comment]
codes add like this:
![image](https://user-images.githubusercontent.com/43697342/132982426-8785b702-8b0c-4a8c-85cd-6121ec4fa20f.png)

**Comment by Flyingdog-Huang** [comment]
maybe I find the reason
![image](https://user-images.githubusercontent.com/43697342/132990502-a9cb72ff-7024-483e-a8bf-30970a4dbb2c.png)

**Comment by k-nayak** [comment]
> Hi,
> 
> For small datasets, reducing the evaluation frequency reduced the training loss for me. This is will avoid learning rate becoming small value after few steps.
> 
> Example.
> 
> ```
> # Evaluation round
> if global_step % (n_train // (0.25 * batch_size)) == 0:
> ```

I have a dataset of 260 images and 0.25 did help significantly to reduce loss, but the DIce Coefficient has remained stagnant at 0.36. Is there any way to improve the DICE score ? It is unable to generalize when given a new image.

**Comment by Flyingdog-Huang** [comment]
I also meet this problem , and I am thinking a way of that

**Comment by Flyingdog-Huang** [comment]
> > Hi,
> > For small datasets, reducing the evaluation frequency reduced the training loss for me. This is will avoid learning rate becoming small value after few steps.
> > Example.
> > ```
> > # Evaluation round
> > if global_step % (n_train // (0.25 * batch_size)) == 0:
> > ```
> 
> I have a dataset of 260 images and 0.25 did help significantly to reduce loss, but the DIce Coefficient has remained stagnant at 0.36. Is there any way to improve the DICE score ? It is unable to generalize when given a new image.

I also meet this problem , and I am thinking a way of that

**Comment by k-nayak** [comment]
Same here. Will update incase I get a fix for better Dice. Please do update if any fix is found. 

Thanks in advance.

**Comment by Flyingdog-Huang** [comment]
> Same here. Will update incase I get a fix for better Dice. Please do update if any fix is found.
> 
> Thanks in advance.

hi, what about your project? 2 class or more class ?

**Comment by k-nayak** [comment]
> > Same here. Will update incase I get a fix for better Dice. Please do update if any fix is found.
> > Thanks in advance.
> 
> hi, what about your project? 2 class or more class ?

mine is binary segmentation

**Comment by Flyingdog-Huang** [comment]
> > > Same here. Will update incase I get a fix for better Dice. Please do update if any fix is found.
> > > Thanks in advance.
> > 
> > 
> > hi, what about your project? 2 class or more class ?
> 
> mine is binary segmentation

oh, that is weird, mine is multi-seg

**Comment by Flyingdog-Huang** [comment]
> > > Same here. Will update incase I get a fix for better Dice. Please do update if any fix is found.
> > > Thanks in advance.
> > 
> > 
> > hi, what about your project? 2 class or more class ?
> 
> mine is binary segmentation

I want to know if the target is smaller than the backgroud in your dataset?
my situation is :
my dataset is small , and target is smaller than backgroud.
in train part, I compute dice loss with backgroud channel, and the loss works well.
in evaluate part, I   compute dice  without backgroud channel , and the dice works bad.
![image](https://user-images.githubusercontent.com/43697342/135198368-bd21ad87-79c0-44cc-a0c8-a8347521c7fa.png)

and I found that the target is almost same as backgroud in this project's data,
so I guess the reason that dice is not good enough is we do not compute the big backgroud channel.
next I will analyze the relationship of dice and big backgroud in math way.

**Comment by k-nayak** [comment]
Very interesting Approach @Flyingdog-Huang. My data set is small as well and the target is smaller compared to the whole image and there are water droplets, which are difficult to distinguish at times and also making masks. Based on the lighting conditions the data tends to be better or worse at times. The model is unable to distinguish at times if a droplet is present or not since it's transparent and lacks robust edges most of the times. I believe with my data set the problem is the the dataset itself. I am implementing it to do real-time segmentation in a video. And the results are not very good. Attention Unet perfomed a little better than U-Net and i am checking If residual Attention U-Net can perform better. My Dice score is around 0.73 with a loss of 0.20 .
